fix(compiler): incorrectly interpreting some HostBinding names

crisbeto opened the pull request:
Currently when analyzing the metadata of a directive, we bundle together the bindings from `host` and the `HostBinding` and `HostListener` together. This can become a problem later on in the compilation pipeline, because we try to evaluate the value of the binding, causing something like `@HostBinding('class.foo') public true = 1;` to be treated the same as `host: {'[class.foo]': 'true'}`.

While looking into the issue, I noticed another one that is closely related: we weren't treating quoted property names correctly. E.g. `@HostBinding('class.foo') public "foo-bar" = 1;` was being interpreted as `classProp('foo', ctx.foo - ctx.bar)` due to the same issue where property names were being evaluated.

These changes resolve both of the issues by treating all `HostBinding` instances as if they're reading the property from `this`. E.g. the `@HostBinding('class.foo') public true = 1;` from above is now being treated as `host: {'[class.foo]': 'this.true'}` which further down the pipeline becomes `classProp('foo', ctx.true)`. This doesn't have any payload size implications for existing code, because we've always been prefixing implicit property reads with `ctx.`. If the property doesn't have an identifier that can be read using dotted access, we convert it to a quoted one (e.g. `classProp('foo', ctx['is-foo']))`.

Fixes #40220.
Fixes #40230.
Fixes #18698.

crisbeto commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -738,7 +738,11 @@ export function extractHostBindings(
+          // through `this` so that further down the line it can't be confused for a literal value
+          // (e.g. if there's a property called `true`). There is no size penalty, because all
+          // values (except literals) are converted to `ctx.propName` eventually.
+          bindings.properties[hostPropertyName] = getSafePropertyAccessString('this', member.name);
```
Note for reviewer: I was conflicted between doing this (converting all accesses to go through `this`) versus changing the extracted information from `Record<string, string>` to `Record<string, {name: string; canBePrimitive: boolean}>` and then changing the places where we evaluate the expression to account for the `canBePrimitive` flag. I decided against it, because it felt more error-prone. I'm open towards changing it, if you think that the `this` access is problematic.

jessicajaniuk replied:
I personally feel that we should avoid `this`. I personally would prefer something more akin to what you mentioned with `Record`, but I can be convinced otherwise.

JoostK replied:
I would also prefer if we could avoid this `this` here. It would probably be better if we were to build an output AST expression instead of synthesizing an expression string, so that we forego the dependency on the binding parser. But I recognize that doing so looks to be easier said than done, as the binding parser implements additional logic that is still relevant for such bindings.

crisbeto replied:
Yeah, my concern is that it'll involve adding a bunch of random IFs everywhere just to handle this one edge case, whereas prefixing it with `this.` reflects what actually happens at runtime. I was looking to change the logic in `BindingParser.parsePropertyBinding` and `Parser.parseSimpleBinding` so that they always return a `PropertyRead`, but getting these classes to know when this should happen would involve adding that extra flag to the extracted information and the data format appears to be used by VE to generate metadata.json files as well.

It's worth noting this isn't the first place where we're constructing the expression as a string like this. If you look a bit further down in the same file, the way we handle `@HostListener` is by creating an expression like `${member.name}(${args.join(',')})` and then evaluating it.

As for building an output AST: I considered doing it, but then I'd also have to bring in an `AbstractEmitterVisitor` which seemed like it would be wasteful, considering that we'd be doing it for each `@HostBinding` and it eventually hits the same `escapeIdentifier` function that I'm using inside `getSafePropertyAccessString`.

petebacondarwin replied:
I actually think that using `this` here is very reasonable.

@JoostK and @jessicajaniuk - you say you prefer not to use it, but did not give the reason for why it is bad.

The only concern I can think of is that we are manually constructing a string expression, but as @crisbeto points out there is already a precedent for this. I feel that `this.` actually states quite clearly what we are trying to achieve here, whereas a more complex approach avoid this string expression would be harder to comprehend and maintain IMO.

JoostK replied:
It's not that I don't like the `this`, it's the creation of an expression string that I'd rather avoid. But I do recognize the difficulty of avoiding it at the moment, so I'd be okay with keeping it as is.

jessicajaniuk replied:
For me, `this` just adds a lot of confusion to follow which `this` scope I may be in at any given time. I'm fine with leaving it as is here given that it seems like any alternative would be way more complicated.

jessicajaniuk commented on packages/compiler/src/render3/util.ts:
```diff
@@ -84,6 +85,11 @@ export function getSyntheticPropertyName(name: string) {
 
+export function getSafePropertyAccessString(accessor: string, name: string): string {
+  const escapedName = escapeIdentifier(name, false, false);
+  return escapedName !== name ? `${accessor}[${escapedName}]` : `${accessor}.${name}`;
```
Do we have to worry about security concerns and execution here? `accessor` and `escapedName` or `name` could literally be anything.

crisbeto replied:
I don't think there's a security concern, because it operates only on the user's source code and the `accessor` is something that we provide.

jessicajaniuk commented on packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_bindings/host_bindings/host_bindings_quoted_names.ts:
```diff
@@ -0,0 +1,12 @@
+
+@Directive({selector: '[hostBindingDir]'})
+export class HostBindingDir {
+  @HostBinding('class.a') 'is-a': any;
```
Is there `any` way around the `any` usage here. Can we make it `unknown` instead.

crisbeto replied:
I figured that it doesn't really matter since it's a test to verify the compiler output. I'll type it better when addressing the rest of the feedback.

jessicajaniuk commented on packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_bindings/host_bindings/host_bindings_quoted_names.js:
```diff
@@ -0,0 +1,11 @@
+
```
(nit) There's a leading space here.

jessicajaniuk commented on packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_bindings/host_bindings/host_bindings_primitive_names.ts:
```diff
@@ -0,0 +1,18 @@
+  }
+})
+export class HostBindingDir {
+  @HostBinding('class.c') true: any;
```
It would be great if we can avoid adding additional `any` types. Can `unknown` work here, instead?

petebacondarwin commented:
@crisbeto - can you mark that this PR will fix #18698 too?

crisbeto committed:
fix(compiler): incorrectly interpreting some HostBinding names

Currently when analyzing the metadata of a directive, we bundle together the bindings from `host`
and the `HostBinding` and `HostListener` together. This can become a problem later on in the
compilation pipeline, because we try to evaluate the value of the binding, causing something like
`@HostBinding('class.foo') public true = 1;` to be treated the same as
`host: {'[class.foo]': 'true'}`.

While looking into the issue, I noticed another one that is closely related: we weren't treating
quoted property names correctly. E.g. `@HostBinding('class.foo') public "foo-bar" = 1;` was being
interpreted as `classProp('foo', ctx.foo - ctx.bar)` due to the same issue where property names
were being evaluated.

These changes resolve both of the issues by treating all `HostBinding` instance as if they're
reading the property from `this`. E.g. the `@HostBinding('class.foo') public true = 1;` from above
is now being treated as `host: {'[class.foo]': 'this.true'}` which further down the pipeline becomes
`classProp('foo', ctx.true)`. This doesn't have any payload size implications for existing code,
because we've always been prefixing implicit property reads with `ctx.`. If the property doesn't
have an identifier that can be read using dotted access, we convert it to a quoted one (e.g.
`classProp('foo', ctx['is-foo']))`.

Fixes #40220.
Fixes #40230.
Fixes #18698.

AndrewKushnir commented:
[Presubmit](http://test/OCL:350465921:BASE:350465939:1609986168474:92fa98e).

atscott commented:
Hi @crisbeto, I changed this to minor only because there is a merge conflict with the 11.0.x branch.
